File: android/rctmgl/src/main/kotlin/com/mapbox/rctmgl/utils/Logger.java
package com.mapbox.rctmgl.utils;
import android.util.Log;


public class Logger {
    public static interface LoggerDefinition {
        public void v(String tag, String msg);
        public void v(String tag, String msg, Throwable tr);
        public void d(String tag, String msg);
        public void d(String tag, String msg, Throwable tr);
        public void i(String tag, String msg);
        public void i(String tag, String msg, Throwable tr);
        public void w(String tag, String msg);
        public void w(String tag, String msg, Throwable tr);
        public void e(String tag, String msg);
        public void e(String tag, String msg, Throwable tr);

    }

    static LoggerDefinition logger = new LoggerDefinition() {
        @Override
        public void v(String tag, String msg) {
            Log.v(tag, msg);
        }

        @Override
        public void v(String tag, String msg, Throwable tr) {
            Log.v(tag, msg, tr);
        }

        @Override
        public void d(String tag, String msg) {
            Log.d(tag, msg);
        }

        @Override
        public void d(String tag, String msg, Throwable tr) {
            Log.d(tag, msg);
        }

        @Override
        public void i(String tag, String msg) {
            Log.i(tag, msg);
        }

        @Override
        public void i(String tag, String msg, Throwable tr) {
            Log.i(tag, msg, tr);
        }

        @Override
        public void w(String tag, String msg) {
            Log.w(tag, msg);
        }

        @Override
        public void w(String tag, String msg, Throwable tr) {
            Log.w(tag, msg);
        }

        @Override
        public void e(String tag, String msg) {
            Log.e(tag, msg);
        }

        @Override
        public void e(String tag, String msg, Throwable tr) {
            Log.e(tag, msg, tr);
        }
    };

    public static void setVerbosity(int logLevel) {
        Logger.logLevel = logLevel;
    }

    public static void setLoggerDefinition(LoggerDefinition dest) {
        logger = dest;
    }

    private static int logLevel;

    static public void d(String tag, String msg) {
        if (logLevel <= Log.DEBUG) {
            logger.d(tag, msg);
        }
    }

    static public void v(String tag, String msg) {
        if (logLevel <= Log.VERBOSE) {
            logger.v(tag, msg);
        }
    }

    static public void i(String tag, String msg) {
        if (logLevel <= Log.INFO) {
            logger.i(tag, msg);
        }
    }

    static public void w(String tag, String msg) {
        if (logLevel <= Log.VERBOSE) {
            logger.w(tag, msg);
        }
    }

    static public void w(String tag,String msg, Throwable tr) {
        if (logLevel <= Log.VERBOSE) {
            logger.w(tag,msg,tr);
        }
    }

    public static void e(String tag, String msg) {
        if (logLevel <= Log.ERROR) {
            logger.e(tag, msg);
        }
    }

    static public void e(String tag, String msg, Throwable tr) {
        if (logLevel <= Log.ERROR) {
            logger.e(tag, msg, tr);
        }
    }


};

File: android/src/main/java/com/rnmapbox/rnmbx/utils/Logger.kt
package com.rnmapbox.rnmbx.utils

import android.util.Log
import com.rnmapbox.rnmbx.utils.Logger.LoggerDefinition

object Logger {
    var logger: LoggerDefinition = object : LoggerDefinition {
        override fun v(tag: String, msg: String) {
            Log.v(tag, msg!!)
        }

        override fun v(tag: String, msg: String, tr: Throwable) {
            Log.v(tag, msg, tr)
        }

        override fun d(tag: String, msg: String) {
            Log.d(tag, msg!!)
        }

        override fun d(tag: String, msg: String, tr: Throwable) {
            Log.d(tag, msg!!)
        }

        override fun i(tag: String, msg: String) {
            Log.i(tag, msg!!)
        }

        override fun i(tag: String, msg: String, tr: Throwable) {
            Log.i(tag, msg, tr)
        }

        override fun w(tag: String, msg: String) {
            Log.w(tag, msg!!)
        }

        override fun w(tag: String, msg: String, tr: Throwable) {
            Log.w(tag, msg!!)
        }

        override fun e(tag: String, msg: String) {
            Log.e(tag, msg!!)
        }

        override fun e(tag: String, msg: String, tr: Throwable) {
            Log.e(tag, msg, tr)
        }
    }

    fun setVerbosity(logLevel: Int) {
        Logger.logLevel = logLevel
    }

    fun setLoggerDefinition(dest: LoggerDefinition) {
        logger = dest
    }

    private var logLevel = 0
    fun d(tag: String, msg: String) {
        if (logLevel <= Log.DEBUG) {
            logger.d(tag, msg)
        }
    }

    fun v(tag: String, msg: String) {
        if (logLevel <= Log.VERBOSE) {
            logger.v(tag, msg)
        }
    }

    fun i(tag: String, msg: String) {
        if (logLevel <= Log.INFO) {
            logger.i(tag, msg)
        }
    }

    fun w(tag: String, msg: String) {
        if (logLevel <= Log.VERBOSE) {
            logger.w(tag, msg)
        }
    }

    fun w(tag: String, msg: String, tr: Throwable) {
        if (logLevel <= Log.VERBOSE) {
            logger.w(tag, msg, tr)
        }
    }

    fun e(tag: String, msg: String) {
        if (logLevel <= Log.ERROR) {
            logger.e(tag, msg)
        }
    }

    @JvmStatic fun e(tag: String, msg: String, tr: Throwable) {
        if (logLevel <= Log.ERROR) {
            logger.e(tag, msg, tr)
        }
    }

    fun logged(tag:String, callback: () -> Unit) {
        try {
            callback()
        } catch (error: Exception) {
            Logger.e(tag, "Exception - error: $error")
        }
    }

    interface LoggerDefinition {
        fun v(tag: String, msg: String)
        fun v(tag: String, msg: String, tr: Throwable)
        fun d(tag: String, msg: String)
        fun d(tag: String, msg: String, tr: Throwable)
        fun i(tag: String, msg: String)
        fun i(tag: String, msg: String, tr: Throwable)
        fun w(tag: String, msg: String)
        fun w(tag: String, msg: String, tr: Throwable)
        fun e(tag: String, msg: String)
        fun e(tag: String, msg: String, tr: Throwable)
    }
}